src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowLoopSpec.groovy

```diff
@@ -0,0 +1,772 @@
+
+        then: "Human readable error is returned" //system can't update the flow when it is down
+        def exc = thrown(HttpClientErrorException)
+        exc.statusCode == HttpStatus.BAD_REQUEST
+        with(exc.responseBodyAsString.to(MessageError)) {
+            errorMessage == "Could not update flow"
+            errorDescription == "Loop switch is not terminating in flow path"
+        }
+
+        then: "FlowLoop is not created"
+        !northbound.getFlow(flow.flowId).loopSwitchId
+
+        and: "FlowLoop rules are not created on the dst switch"
+        getFlowLoopRules(transitSwId).empty
+
+        cleanup:
+        flow && flowHelperV2.deleteFlow(flow.flowId)
+    }
+
+    @Tidy
+    @Tags(LOW_PRIORITY)
+    def "Unable to create flowLoop for a non existent flow"() {
+        when: "Try to create flowLoop on the transit switch"
+        def sw = topology.activeSwitches.first()
+        northboundV2.createFlowLoop(NON_EXISTENT_FLOW_ID, new FlowLoopPayload(sw.dpId))
+
+        then: "Human readable error is returned"
+        def exc = thrown(HttpClientErrorException)
+        exc.statusCode == HttpStatus.NOT_FOUND
+        with(exc.responseBodyAsString.to(MessageError)) {
+            errorMessage == "Flow not found"
+            errorDescription == "Flow '$NON_EXISTENT_FLOW_ID' not found."
+        }
+
+        and: "FlowLoop rules are not created on the switch"
+        getFlowLoopRules(sw.dpId).empty
+
+        and: "The switch is valid"
+        northbound.validateSwitch(sw.dpId).verifyRuleSectionsAreEmpty(["missing", "excess", "misconfigured"])
+        def switchIsValid = true
+
+        cleanup:
+        !switchIsValid && northbound.synchronizeSwitch(sw.dpId, true)
+    }
+
+    @Tidy
+    @Tags(LOW_PRIORITY)
+    def "Unable to create flowLoop on a non existent switch"() {
+        given: "An active multi switch flow"
+        def swP = topologyHelper.getNeighboringSwitchPair()
+        def flow = flowHelperV2.randomFlow(swP)
+        flowHelperV2.addFlow(flow)
+
+        when: "Try to create flowLoop on a non existent switch"
+        northboundV2.createFlowLoop(flow.flowId, new FlowLoopPayload(NON_EXISTENT_SWITCH_ID))
+
+        then: "Human readable error is returned"
+        def exc = thrown(HttpClientErrorException)
+        exc.statusCode == HttpStatus.BAD_REQUEST
+        with(exc.responseBodyAsString.to(MessageError)) {
+            errorMessage == "Could not update flow"
+            errorDescription == "Loop switch is not terminating in flow path"
+        }
+
+        and: "FlowLoop rules are not created for the flow"
+        !northboundV2.getFlow(flow.flowId).loopSwitchId
+
+        cleanup:
+        flow && flowHelperV2.deleteFlow(flow.flowId)
+    }
+
+    def "getSwPairConnectedToTraffGenForSimpleFlow"(List<SwitchId> switchIds) {
+        getTopologyHelper().getAllNotNeighboringSwitchPairs().collectMany { [it, it.reversed] }.find { swP ->
```

Comment:
`getAllNotNeighboringSwitchPairs` still not fixed here and in all methods below. Is it possible to just use `getSwitchPairs()` and operate with all available topology?